pretyping/find_subterm.ml

```diff
@@ -58,55 +58,87 @@ let map_named_declaration_with_hyploc f hyploc acc decl =
 
 exception SubtermUnificationError of subterm_unification_error
 
-exception NotUnifiable of (EConstr.t * EConstr.t * unification_error) option
-
 type 'a testing_function = {
-  match_fun : 'a -> EConstr.constr -> 'a;
-  merge_fun : 'a -> 'a -> 'a;
+  match_fun : 'a -> EConstr.constr -> ('a, unit) Result.t;
```

Comment:
Why not 'a option?

Reply:
Some intermediate state required a payload. Also, the intent is clearer to the user.

Reply:
The intermediate state doesn't matter.

>the intent is clearer to the user.

No it's not, we use option everywhere else for this sort of thing.

Reply:
Actually, the overreliance on option is a bit counterproductive at times. We use it for historical reasons, because the result types has only been around for a few OCaml versions (since 4.08 it seems). Option types should be used when the function has to return a dummy value, while result should be used to signal error instead. The fact that the error is dummy is not really relevant I think.

But if you insist I'll turn this into an option type...

Reply:
Actually I did find some dead code issues thanks to the use of result rather than option, I think it'd be better to keep it that way in the end.